### ISSUE TITLE
The Big Refactor

### ISSUE THREAD

**Issue by mdpcs3544** [CharlotteOS-Deprecated/charlotte-core#135]
Refactor the entire kernel codebase to accomplish the following:

- [x] Split the kernel cleanly into subsystems that each do one thing only and do it well.

- [ ] Break up arch so that each subsystem has its own isa module within it to encapsulate any ISA specific code needed by that subsystem.

- [ ] Only the top level of each subsystem should expose any pub items but it can have pub modules so long as the pub items in them are reexported as part of the top level module. 

    - That would look like this:

```
    pub mod my_mod;
    use my_mod::*;
```
  

- [ ] All top level modules' public APIs should be Rust extern "C" ABI compatible.

**Comment by Wrench56** [comment]
One more thing: I would also like to eliminate all Clippy warnings. Yep, I did do this on #101 but it became problematic with the `pmm` + `acpi` merges. So I think the easiest way to fix it is to simply redo it.

**Comment by mdpcs3544** [comment]
> One more thing: I would also like to eliminate all Clippy warnings. Yep, I did do this on #101 but it became problematic with the `pmm` + `acpi` merges. So I think the easiest way to fix it is to simply redo it.

I think this warrants its own separate issue after this one is complete. Feel free to write it up and list it as blocked on this issue.

**Comment by mdpcs3544** [comment]
This is way too much work and would almost require a rewrite of most of the kernel. We can't really afford to make changes this large. Thus this is being cancelled and replaced with a task to refactor arch into one single trait and a set of C wrappers in the bridge library.
